# pyCfS
The aggregation of Lichtarge Lab genotype-phenotype validation experiments

#### Installation
pip install git+https://github.com/kevwilhelm95/pyCfS.git

#### Examples
See "example.ipynb" for help

#### Parallelization notes
Parallelized functions require the user to run function under blocking guard (i.e. if __name__ == "__main__":)

## Modules
### pyCFS.Combine
|   Function    |   Required arguments |   Optional parameters |   Returns  |   Parallelized?   |
|---------------|----------------------|-----------------------|------------|-------------------|
|consensus()    |1. genes_1 (list): list of genes<br>2. genes_2 (list): list of genes| 1. genes_3 (list)<br>2. genes_4 (list)<br>3. genes_5 (list)<br>4. genes_6 (list)<br>5. list_names (list): names of your lists (Default set to "set_x")<br>6. plot_fontface (str): Font for upset plot (Default = Avenir)<br>7. plot_fontsize (int): Font size for upset plot (Default = 14)<br>8. savepath (str): Path to save files. If no path provided, files are not saved.| 1. pd.DataFrame: A dataframe with columns 'gene', 'occurrences', and 'lists' detailing each unique gene, the number of its occurrences, and the lists it appeared in.<br>2. Image: Upset plot showing overlap between input genelists.| No|
|functional_clustering | 1. genes_1 (list): list of genes<br>2. genes_2 (list): list of genes | 1. genes_3 (list)<br>2. genes_4 (list)<br>3. genes_5 (list)<br>4. source_names (list): names of the gene lists<br>5. evidences (list): Evidences for computing edge weight (Default = ['all']. Components = [neighborhood, fusion, cooccurence, coexpression, experimental, database, textmining]).<br>6. edge_confidence (str): Minimum edge weight for network (Default = 'highest'. Options = all (>0), low (>0.2), medium (>0.4), high (>0.7), highest (>0.9)).<br>7. random_iter (int): Number of random iterations to perform for background (Default = 100)<br>8. inflation (float): Set inflation parameter. If not set, algorithm will determine optimal inflation from 1.5-3.0.<br>9.pathways_min_group_size (int): Minimum group size for functional enrichment (Default = 5).<br>10. pathways_max_group_size (int): Max group size for functional enrichment (Default = 100).<br>11. cores (int): Number of cores for parallelization (Default = 1).<br>11. savepath (str): Path to save. If not provided, no files will be saved.| 1. pd.DataFrame: Pairwise edges for true connection network.<br>2. pd.DataFrame: Table of genes, clusters, and their source.<br>3. dict: Diction of dataframes by cluster containing functional enrichment. | Yes|
| statistical_combination | 1. df_1 (pd.DataFrame): two-column df with genes in column 1 and p-value in column 2. Headers do not matter.<br>2. df_2 (pd.DataFrame): two-column df with genes in column 1 and p-value in column 2. | 1. df_3 (pd.DataFrame)<br>2. df_4 (pd.DataFrame)<br>3. df_5 (pd.DataFrame)<br>4. df_6 (pd.DataFrame)<br>5. savepath (str): Path to save. No files saved if not provided.| 1. pd.DataFrame: Dataframe containing genes, df_x p-values, and the combinations for Cauchy, MinP, CMC, MCM, and p-values multiplied.| No|

### pyCFS.GoldStandards
|   Function    |   Required arguments |   Optional parameters |   Returns  |   Parallelized?   |
|---------------|----------------------|-----------------------|------------|-------------------|
| goldstandard_overlap| 1. query(list): list of query genes<br>2. goldstandard(list): list of goldstandard genes | 1. plot_query_color(str): Color of venn diagram (Default = red)<br>2. plot_goldstandard_color (str): Color of venn diagram (Default = gray)<br>3. plot_fontsize (int): Fontsize for venn diagram (Default = 14)<br>4. plot_fontface (str): Fontface for Venn Diagram (Default = Avenir)<br>5. savepath (str): Path for saving. If not provided, no files are saved.|1. list: list of overlapping genes.<br>2. float: hypergeometric pvalue of overlap<br>3. image: Venn Diagram of overlap.| No|
|ndiffusion | Coming | Coming | Coming | Coming|
|interconnectivity | 1. set_1 (list): list of genes.<br>2. set_2 (list): list of genes.| 1. set_3 (list)<br>2. set_4 (list)<br>3. set_5 (list)<br>3. evidences (list): Evidences for computing edge weight (Default = ['all']. Components = [neighborhood, fusion, cooccurence, coexpression, experimental, database, textmining]).<br>4. edge_confidence (str): Minimum edge weight for network (Default = 'highest'. Options = all (>0), low (>0.2), medium (>0.4), high (>0.7), highest (>0.9)).<br>5. num_iterations (int): # of iterations for background connections (Default = 250).<br>6. cores (int): cores for parallelization (Default = 1).<br>7. plot_fontface (str): Default = Avenir.<br>8. plot_fontsize (int): Default = 14.<br>9. plot_background_color (str): Histogram color for enrichment (Default = gray).<br>10. plot_query_color (str): Line color for enrichment (Default = red).<br>11. savepath (str).| 1. Image: Venn diagram of overlap.<br>2. Image: Hisogram of randomized connections v. query connections.<br>3. list: list of random connections.<br>4. pd.DataFrame: Unique gene network.<br>5. dict: gene_sources.|Yes|
|gwas_catalog_colocalization | 1. query (list): list of genes. | 1. mondo_id (str): ID from GWAS Catalog of disease. If used, calls API.<br>2. gwas_summary_path (str): Path to GWAS Catalog downloaded summary statistics. If used, API is not called.<br>3. gwas_p_thresh (float): Filter for GWAS Catalog loci to compare to. (Default = 5e-8).<br>4. distance_mbp (float): Distance threshold for colocalization in Mbp (Default = 0.5 Mbp).<br>5. cores (int): # of cores for parallelized background (Default = 1).<br>6. savepath (str): Path to save.<br>7. save_summary_statistics (bool): Save downloaded summary stats if savepath is also defined.| 1. pd.DataFrame: ['Gene', 'SNP'] detailing snps that colocalize with a gene.<br>2. float: Fisher's exact test of colocalization | Yes |
|pubmed_comentions | 1. query (list): List of genes.<br>2. keyword (str): Keyword to search co-mentions for. | 1. email (str): Email of PubMed account (Default is my extended account)<br>2. api_key (str): API key of PubMed account.<br>3. enrichment_trials (int): # of randomiztion trials (Default = 100).<br>4. workers (int): # of workers for threaded API calls (Default = 15).<br>5. run_enrichment (bool): Toggle running background enrichment (Default  True).<br>6. enrichment_cutoffs (list): Breaks for determining enrichment (>1st number, <=2nd number; Default = [[-1,0], [0,5], [5,15], [15,50], [50,10000]]).<br>7. plot_background_color (str): Color for background distribution (Default = gray).<br>8. plot_query_color (str): Color for true observation (Default = red).<br>8. plot_fontface (str): Default = Avenir.<br>9. plot_fontsize (int): Default = 14.<br>10. savepath (str): Path to save | 1. pd.DataFrame: Dataframe of genes, co-mention counts, and PMIDs.<br>2. dict: Keys = enrichment_cutoff values; Values = (# of query genes, z_score).<br>3. dict: Keys = Enrichment_cutoff values; Values = Image of enrichment distribution | Yes |

### pyCFS.Clinical
|   Function    |   Required arguments |   Optional parameters |   Returns  |   Parallelized?   |
|---------------|----------------------|-----------------------|------------|-------------------|
| mouse_phenotype_enrichment |1. query (list): List of genes | 1. background (str): Background gene set (Default = ensembl. Options = ensembl, Reactomes)<br>2. random_iter (int): Iterations for background runs (Default = 5000).<br>3. plot_sig_color (str): Color for strip_plot significant genes (Default = red).<br>4. plot_q_threshold (float): Significance threshold for strip_plot (Default = 0.05).<br>5. plot_show_labels (bool): Choose to plot labels. If true, must also provide labels in plot_labels_to_show (Default = False).<br>6. plot_labels_to_show (list): Phenotype labels to plot. Use "modelPhenotypeLabel" in output dataframe.<br>7. plot_fontface (str): Default = Avenir.<br>8. plot_fontsize (int):Default = 14.<br>9. cores (int): Default = 1.<br>10. savepath (str): Path to save. No files saved if empty. | 1. pd.DataFrame: Summary of enrichments.<br>2. Image: Strip plot of enrichment.<br>3. Image: Histogram of FDR values. | Yes|
| protein_family_enrichment | Coming | Coming | Coming | Coming |
| tissue_expression_enrichment | Coming | Coming | Coming | Coming|
| depmap_enrichment | Coming | Coming | Coming | Coming|
| drug_targets | Coming | Coming | Coming | Coming |
